watch/watch.go

```diff
@@ -0,0 +1,360 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package watch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/storage"
+)
+
+// Watcher allows for watches to be registed on queries. When a dependency of a watched
+// query changes, a notification will be send back to the user.
+type Watcher struct {
```

Comment:
Watcher is responsible for both the watch itself (i) and the management of watches (ii). I'm not sure if this is intentional or not. I would consider separating (i) into another struct (e.g., `Watch`) and use Watcher solely for management (e.g., creation and deletion) of watches (following the single responsibility principle.